key-manager/src/lib.rs

```diff
@@ -0,0 +1,344 @@
+    pub epoch: u64,
+    pub key: Secret<Aes256GcmDiskEncryptionKey>,
+}
+
+/// The main mechanism used to derive keys from a shared secret for the Oxide
+/// control plane.
+pub struct KeyManager<S: SecretRetriever> {
+    // A mechanism for retrieving input key material
+    secret_retriever: S,
+
+    /// Pseudo-Random-Keys (PRKs) wrapped in an `Hkdf` structure so that we can
+    /// create keys from them with `HKDF-Expand` (`prk.expand_multi_info`).
+    ///
+    /// In the common case, we will only have a single PRK for everything.
+    ///
+    /// If there is an ongoing reconfiguration, we will have at least 2 PRKs. In
+    /// some cases of failure while multiple reconfigurations have taken place,
+    /// we may have multiple PRKs. Policy should limit the number of allowed
+    /// reconfigurations before a node is no longer part of the trust quorum.
+    /// This should most likely be a small number like `3`.
+    prks: BTreeMap<u64, Hkdf<Sha3_256>>,
+}
+
+impl<S: SecretRetriever> KeyManager<S> {
+    pub fn new(secret_retriever: S) -> KeyManager<S> {
+        KeyManager { secret_retriever, prks: BTreeMap::new() }
+    }
+
+    /// Load latest version of the input key material into the key manager.
+    pub async fn load_latest_secret(&mut self) -> Result<(), Error> {
+        let ikm = self.secret_retriever.get_latest().await?;
+        self.insert_prk(ikm);
+        Ok(())
+    }
+
+    /// Load input key material for the given epoch into the key manager.
+    pub async fn load_secret(&mut self, epoch: u64) -> Result<(), Error> {
+        match self.secret_retriever.get(epoch).await? {
+            SecretState::Current(ikm) => self.insert_prk(ikm),
+            SecretState::Reconfiguration { old, new } => {
+                self.insert_prk(old);
+                self.insert_prk(new);
+            }
+        }
+        Ok(())
+    }
+
+    /// Derive an encryption key for the given [`sled_hardware::DiskIdentity`]
+    pub async fn disk_encryption_key(
+        &mut self,
+        epoch: u64,
+        disk_id: &DiskIdentity,
+    ) -> Result<VersionedAes256GcmDiskEncryptionKey, Error> {
+        let prk = if let Some(prk) = self.prks.get(&epoch) {
+            prk
+        } else {
+            self.load_secret(epoch).await?;
+            self.prks.get(&epoch).unwrap()
+        };
+
+        let mut key = Aes256GcmDiskEncryptionKey::default();
+
+        // Unwrap is safe because we know our buffer is large enough to hold the output key
+        prk.expand_multi_info(
+            &[
+                b"U.2-zfs-",
+                disk_id.vendor.as_bytes(),
+                disk_id.model.as_bytes(),
+                disk_id.serial.as_bytes(),
+            ],
+            key.0.as_mut(),
+        )
+        .unwrap();
+
+        Ok(VersionedAes256GcmDiskEncryptionKey { epoch, key: Secret::new(key) })
+    }
+
+    /// Return the epochs for all secrets which are loaded
+    pub fn loaded_secrets(&self) -> Vec<u64> {
```

Comment:
Nit - this method name sounds like it returns the secrets themselves. `epochs_of_loaded_secrets()` is a mouthful - maybe just `epochs()`?

Reply:
Good point. I'll rename. `loaded_epochs` is probably a good middle ground.

Reply:
Fixed in [75d92cf](https://github.com/oxidecomputer/omicron/pull/2990/commits/75d92cf2095a16ce622234f5c06c16d232e4a712)